docs/source/api-life-cycle.md

```diff
@@ -0,0 +1,225 @@
+4. Once the deprecation period has passed, the API may be removed (See
+   “Deprecation Period” below). Be sure to also remove references from the
+   documentation.
+
+
+We also use deprecation as a way to make breaking changes to an existing
+interface: for example, if adding a non-optional parameter to a method. To do
+this without breaking existing users:
+
+1. In a single commit:
+   - Create a new API that meets the new requirements.
+   - Deprecate the old API and recommend that users move to the new API.
+2. Migrate use cases from the old API to the new API.
+3. Delete the old API after the deprecation period.
+
+## How to Mark API State
+
+When possible, the ExecuTorch code uses language-standard ways to annotate API
+lifecycle state in the code. This makes it easier for IDEs and other tools to
+communicate state to developers.
+
+<table>
+  <tr>
+   <td><strong>Language</strong>
+   </td>
+   <td><strong>Code</strong>
+   </td>
+   <td><strong>Documentation</strong>
+   </td>
+  </tr>
+  <tr>
+   <td>Python
+   </td>
+   <td>
+
+Use the
+<a href="https://typing-extensions.readthedocs.io/en/latest/#typing_extensions.deprecated">typing_extensions.deprecated</a>
+decorator
+
+<p>
+Use ExecuTorch's native experimental decorator (TODO not yet implemented)
+
+   </td>
+   <td>
+
+Use <code>.. warning::</code> in the docstrings of deprecated and experimental
+APIs. See
+<a href="https://github.com/pytorch/pytorch/blob/cd8bbdc71a0258292381a7d54c8b353988d02ff4/torch/nn/utils/stateless.py#L170">example
+usage</a>
+
+</ul>
+   </td>
+  </tr>
+  <tr>
+   <td>C++
+   </td>
+   <td>
+
+Use <code>__ET_DEPRECATED</code> macros. See <a href="https://github.com/pytorch/executorch/blob/8e0f856ee269b319ac4195509cf31e3f548aa0e8/runtime/executor/program.h#L81">example usage</a>
+
+<p>
+<p>
+Use <code>__ET_EXPERIMENTAL</code> macros (TODO not yet implemented)
+</ul>
+   </td>
+   <td>
+
+Start Doxygen comments with <code>DEPRECATED:</code> See
+<a href="https://github.com/pytorch/executorch/blob/9d859653ae916d0a72f6b2b5c5925bed38832140/runtime/executor/program.h#L139">example
+usage</a>
+
+<p>
+<p>
+Start Doxygen comments with <code>EXPERIMENTAL:</code>
+   </td>
+  </tr>
+  <tr>
+   <td>Java
+   </td>
+   <td>
+
+Use <a href="https://docs.oracle.com/javase/9/docs/api/java/lang/Deprecated.html">java.lang.Deprecated</a>
+
+<p>
+<p>
+
+Use <a href="https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:docs/api_guidelines/annotations.md">androidx.annotation.RequiresOptIn</a>
+
+   </td>
+   <td>
+<p>
+   <code>/**
+<p>
+* @deprecated Use {@link #newMethod()} instead.
+<p>
+*/
+<p>
+</code>
+<p>
+<code>
+/**
+<p>
+* Warning: This API is experimental.
+<p>
+*/</code>
+   </td>
+  </tr>
+  <tr>
+   <td>Objective-C
+   </td>
+   <td>
+<p>
+<code>__attribute__((deprecated("Use newMethod instead")));</code>
+<p>
+<p>
+<code>__attribute__((experimental("Use newMethod instead")));</code> (TODO not yet implemented)
```

Comment:
I don't know if it's possible to implement an attribute like this; they're typically defined by the toolchain.

Reply:
I'll let @shoumikhin comment on this. I've chatted with him offline and IIUC, there's a mapping between the Objective-C annotations and the CPP macros. So if the CPP `__ET_DEPRECATED` and `__ET_EXPERIMENTAL` can be implemented, then we can implement these.

Reply:
Oh I meant we can use the deprecated annotation for experental features too, since there's no dedicated experimental annotation available. Thus __ET_EXPERIMENTAL is gonna be defined the same as deprecated, but with a different error message.